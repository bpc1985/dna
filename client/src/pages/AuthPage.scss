@import "stylesheets/index";

.authForm {
  width: 25rem;
  max-width: 80%;
  margin: 5rem auto;

  .formControl {
    margin-bottom: 1rem;

    label {
      width: 100%;
      display: block;
      margin-bottom: 0.5rem;
    }

    input {
      padding: 5px 10px;
      margin: 8px 0;
      width: 100%;
      display: block;
      border: 1px solid $light-gray;
      box-sizing: border-box;
    }

    .errorMessage {
      color: $pink;
    }
  }

  .formActions {
    button {
      background: $pink;
      font: inherit;
      border: 1px solid $pink;
      border-radius: 3px;
      padding: 0.25rem 1rem;
      margin-right: 1rem;
      box-shadow: 1px 1px 5px rgba(0, 0, 0, 0.26);
      color: $white;
      cursor: pointer;

      &:hover, &:active {
        background: $purple;
        border-color: $purple;
      }
    }
  }
}