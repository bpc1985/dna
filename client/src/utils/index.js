const SERVER_URL = "http://localhost:8081/";

export const fetchData = async path => {
  const res = await fetch(SERVER_URL + path);
  return await res.json();
};

export const isSearchTermIncluded = (dnaPackage, searchTerm) => {
  return dnaPackage.name.toLowerCase().includes(searchTerm.toLowerCase())
    || dnaPackage.type.toLowerCase().includes(searchTerm.toLowerCase())
    || dnaPackage.agreement.toLowerCase().includes(searchTerm.toLowerCase());
};
