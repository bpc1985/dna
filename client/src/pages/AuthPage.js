import React, { useState } from "react";
import * as Yup from "yup";
import { Formik, Field, Form, ErrorMessage } from "formik";
import { postLogIn, getUrlParams } from "../utils";

import "./AuthPage.scss";

export default function AuthPage(props) {
  const params = getUrlParams(props.location.search);
  const [status, setStatus] = useState(true);

  const submitHandler = async (values) => {
    try {
      const res = await postLogIn({
        email: values.email,
        password: values.password
      });
      localStorage.setItem('token', res.token);
      props.userHasAuthenticated(true);
      props.history.push(params.redirect || '/');
    } catch (e) {
      setStatus(false);
    }
  };

  return (
    <Formik
      initialValues={{
        email: "",
        password: ""
      }}
      validationSchema={Yup.object({
        email: Yup.string()
          .email("Email should be in right format")
          .required("Email is required"),
        password: Yup.string()
          .min(7, "Password must be 7 characters or more")
          .required("Password is required")
      })}
      onSubmit={(values, { setSubmitting }) => {
        submitHandler(values);
        setSubmitting(false);
      }}
    >
      <Form className="authForm">
        {!status && <div className="formControl">
          <div className="errorMessage">Ops! Something is wrong with your email or password</div>
        </div>}
        <div className="formControl">
          <label htmlFor="email">Email</label>
          <Field id="email" name="email" type="text" />
          <ErrorMessage component="div" name="email" className="errorMessage" data-testid="emailError" />
        </div>
        <div className="formControl">
          <label htmlFor="password">Password</label>
          <Field id="password" name="password" type="password" />
          <ErrorMessage component="div" name="password" className="errorMessage" data-testid="passwordError" />
        </div>
        <div className="formActions">
          <button type="submit">Submit</button>
        </div>
      </Form>
    </Formik>
  );
}
