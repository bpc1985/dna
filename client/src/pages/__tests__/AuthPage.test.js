import { render, fireEvent, wait } from "@testing-library/react";
import React from "react";
import AuthPage from "../AuthPage";
import { postLogIn } from "../../utils";

jest.mock("../../utils", () => ({
  postLogIn: jest.fn(() => ({token: 'newToken'})),
  getUrlParams: () => ({
    redirect: "/subscriptions"
  })
}));

const props = {
  location: {
    search: "login?redirect=/subscriptions"
  },
  history: {
    push: jest.fn()
  },
  userHasAuthenticated: jest.fn()
};

describe("AuthPage", () => {
  it("should render component", () => {
    const wrapper = render(<AuthPage {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should be able validate form", async () => {
    const { getByLabelText, getByTestId } = render(<AuthPage {...props} />);

    const emailInput = getByLabelText("Email");
    fireEvent.blur(emailInput);

    await wait(() => {
      expect(getByTestId("emailError")).not.toBe(null);
      expect(getByTestId("emailError")).toHaveTextContent("Email is required");
    });

    fireEvent.change(emailInput, { target: { value: "wrongEmail" } });
    await wait(() => {
      expect(getByTestId("emailError")).toHaveTextContent(
        "Email should be in right format"
      );
    });

    const passwordInput = getByLabelText("Password");
    fireEvent.blur(passwordInput);

    await wait(() => {
      expect(getByTestId("passwordError")).not.toBe(null);
      expect(getByTestId("passwordError")).toHaveTextContent(
        "Password is required"
      );
    });

    fireEvent.change(passwordInput, { target: { value: "12345" } });
    await wait(() => {
      expect(getByTestId("passwordError")).toHaveTextContent(
        "Password must be 7 characters or more"
      );
    });
  });

  it("should be able submit form", async () => {
    const { container, getByLabelText } = render(
      <AuthPage
        location={props.location}
        history={props.history}
        userHasAuthenticated={props.userHasAuthenticated}
      />
    );

    const emailInput = getByLabelText("Email");
    const passwordInput = getByLabelText("Password");
    const loginButton = container.querySelector("button");

    fireEvent.change(emailInput, { target: { value: "user1@mail.com" } });
    fireEvent.change(passwordInput, { target: { value: "abcd@1234" } });
    fireEvent.click(loginButton);

    await wait(() => {
      expect(postLogIn).toHaveBeenCalled();
      expect(localStorage.setItem).toHaveBeenLastCalledWith("token", "newToken");
      expect(props.history.push).toHaveBeenCalled();
      expect(props.userHasAuthenticated).toHaveBeenCalled();
    });
  });
});
