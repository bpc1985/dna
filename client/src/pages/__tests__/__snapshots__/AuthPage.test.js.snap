// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthPage should render component 1`] = `
<div>
  <form
    action="#"
    class="authForm"
  >
    <div
      class="formControl"
    >
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <div
      class="formControl"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="formActions"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
